src/gpl/test/gpl_aux.py

```diff
@@ -0,0 +1,195 @@
+import utl
+
+# Definitions
+DEFAULT_TARGET_DENSITY = 0.7
+
+
+# Besides design, there are no positional args here. General strategy is that 
+# when an arg is None, we just skip setting it, otherwise we will set the
+# parameter after a quick type check.
+def global_placement(design, *,
+    skip_initial_place=False,
+    skip_nesterov_place=False,
+    timing_driven=False,
+    routability_driven=False,
+    incremental=False,
+    force_cpu=False,
+    skip_io=False,
+    bin_grid_count=None,       # positive int, default 0
+    density=None,              # 'uniform' or 0.0 < d < 1.0  default 0.7
+    init_density_penalty=None, # positive float  default 0.00008
+    init_wirelength_coef=None, # positive float  default 0.25
+    min_phi_coef=None,         # positive float  default 0.95
+    max_phi_coef=None,         # positive float  default 1.05
+    reference_hpwl=None,       # positive int    default 446000000
+    overflow=None,             # positive float
+    initial_place_max_iter=None,     # positive int, default 20
+    initial_place_max_fanout=None,   # positive int, default 200
+    routability_check_overflow=None, # positive float
+    routability_max_density=None,    # positive float  default  0.99
+    routability_max_bloat_iter=None, # positive int  default  1
+    routability_max_inflation_iter=None,      # positive int  default 4
+    routability_target_rc_metric=None,        # positive float
+    routability_inflation_ratio_coef=None,    # positive float
+    routability_max_inflation_ratio=None,     # positive float
+    routability_rc_coefficients=None,         # a list of four floats
+    timing_driven_net_reweight_overflow=None, # list of ints
+    timing_driven_net_weight_max=None,        # float
+    timing_driven_nets_percentage=None,       # float
+    pad_left=None, # positive int
+    pad_right=None # positive int
+):
+    gpl = design.getReplace()
+
+    if skip_initial_place:
+        gpl.setInitialPlaceMaxIter(0)
+    elif is_pos_int(initial_place_max_iter):
+        gpl.setInitialPlaceMaxIter(initial_place_max_iter)
+
+    gpl.setForceCPU(force_cpu)
+
+    gpl.setSkipIoMode(skip_io)
+    if skip_io:
+        gpl.setInitialPlaceMaxIter(0) 
+
+    gpl.setTimingDrivenMode(timing_driven)
+
+    if timing_driven:
+        if design.evalTclString('get_libs -quiet "*"') == '':
+            utl.error(utl.GPL, 502, "No liberty libraries found.")
+
+        if skip_io:
+            utl.warn(utl.GPL, 503, "-skip_io will disable timing driven mode.")
+            gpl.setTimingDrivenMode(False)
+
+        if timing_driven_net_reweight_overflow != None:
+            overflow_list = timing_driven_net_reweight_overflow
+        else:
+            overflow_list = [79, 64, 49, 29, 21, 15]
```

Comment:
Why is this repeated here?  Can't we let the c++ defaults be used instead.

Reply:
Didn't know these were set in C++ and just ported the tcl to Python. I think it is best, as you are suggesting, to take them out of the Python code so that changed defaults in C++ will propagate to the Python api. But, tcl users will be insulated from those changes.

Reply:
Annoyingly tcl set the defaults in the command to different values than defaults in c++.  I guess the c++ default values are unused in the current setup.  I guess this matches and is therefore ok for now.  Really we need both to stop setting the default and migrate the value to c++

Reply:
This doesn't appear to be in the C++ code, it is only in the replace.tcl code.

Reply:
I thought it was, perhaps it moved or I mis-remember.